### PR TITLE
Organize sensor fusion

### PR DESCRIPTION
**Pull request by antimonyB**
Moved Mahony files and refactored code.

### CODE REVIEW

File: Autopilot/Inc/gps.hpp [hunk 1]
```diff
@@ -55,7 +55,7 @@ class Gps
 		* or with old data (in case of old data, the dataIsNew flag of the result struct will be cleared).
 		* @param[out]		Data 		pointer to the result struct.
 		*/
-        virtual void GetResult(GpsData_t *Data) = 0;
+        virtual void GetResult(GpsData_t &Data) = 0;
```
Comment (Pop0097) [inline Autopilot/Inc/gps.hpp]: pls fix to match style guide: https://uwarg-docs.atlassian.net/wiki/spaces/ZP/pages/1827635203/Style+Guide

File: Autopilot/Src/gps.cpp [hunk 2]
```diff
@@ -424,27 +424,27 @@ void NEOM8::parse_gga(uint8_t* data) {
 	gpsData.altitude = (int)(tAltitude / multiplier);
 }
 
-void NEOM8::GetResult(GpsData_t * Data) {
+void NEOM8::GetResult(GpsData_t &Data) {
```
Comment (Pop0097) [inline Autopilot/Src/gps.cpp]: style guide: https://uwarg-docs.atlassian.net/wiki/spaces/ZP/pages/1827635203/Style+Guide

`GpsData_t& Data`

File: Autopilot/SensorFusion/Src/SensorFusion.cpp [hunk 3]
```diff
@@ -126,7 +126,7 @@ double * cartesianToGPS(double x, double y){
     return latLong;
 }
 
-SFError_t SF_GetAttitude(SFAttitudeOutput_t *Output, IMUData_t *imudata, airspeedData_t *airspeeddata) {
+SFError_t SF_GetAttitude(SFAttitudeOutput_t *Output, IMUData_t *imudata) {
```
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp:129]: There a reason we are remving the airspeed data? Or is it not needed anymore

**Review by Pop0097 — CHANGES_REQUESTED** [review]
Style guide stuff mostly

Would it also be possible to define the functions for getting sensor data for airspeed, etc? I don't think they've been done

File: Autopilot/SensorFusion/Src/SensorFusion.cpp (hunk 3 above)
Comment (antimonyB) [reply Autopilot/SensorFusion/Src/SensorFusion.cpp:129]: The original attitude function included airspeed but never actually used the data, so it's not needed.

Comment (Pop0097) [reply Autopilot/SensorFusion/Src/SensorFusion.cpp:129]: sounds good

**Review by Pop0097 — COMMENTED** [review]
Looks good to me. Just wanted to confirm how your SF tests performed locally and then I can approve

**Comment by antimonyB** [comment]
After these changes I don't see any issues with the SF tests.

**Comment by Pop0097** [comment]
Awesome! We will merge this in after the S2021 branch is merged in!

File: Autopilot/CMakeLists.txt [hunk 4]
```diff
@@ -246,13 +248,15 @@ set(SENSOR_FUSION_MODULES_SOURCES
 
   set(SENSOR_FUSION_MODULES_UNIT_TEST_SOURCES
     ${CMAKE_CURRENT_SOURCE_DIR}/Test/Src/SensorFusion/Test_SF.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/AttitudeManager/Src/MahonyAHRS.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/SensorFusion/Src/MahonyAHRS.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/CControl/Src/mul.c
     ${CMAKE_CURRENT_SOURCE_DIR}/CControl/Src/tran.c
     ${CMAKE_CURRENT_SOURCE_DIR}/CControl/Src/inv.c
     ${CMAKE_CURRENT_SOURCE_DIR}/CControl/Src/lup.c
```
Comment (Pop0097) [inline Autopilot/CMakeLists.txt]: All of this (lines 251 to 255) should be in `SENSOR_FUSION_MODULES_SOURCES`, not `SENSOR_FUSION_MODULES_UNIT_TEST_SOURCES`

`SENSOR_FUSION_MODULES_UNIT_TEST_SOURCES` should only contain the university ttest files. 

Same applies for the attitude manager tests on line 173 and 174.

File: Autopilot/SensorFusion/Src/SensorFusion.cpp [hunk 5]
```diff
@@ -560,3 +551,44 @@ Airspeed_Data_t SF_GetRawAirspeed()
...
+    gpsOutput.latitude = gpsData.latitude;
+    gpsOutput.longitude = gpsData.longitude;
+    gpsOutput.utcTime = gpsData.utcTime;
+    gpsOutput.groundSpeed = gpsData.groundSpeed;
+    gpsOutput.altitude = gpsData.altitude;
+    gpsOutput.heading = gpsData.heading;
+    gpsOutput.numSatellites = gpsData.numSatellites;
+    gpsOutput.fixStatus = gpsData.fixStatus;
+    gpsOutput.sensorStatus = gpsData.sensorStatus;
+    gpsOutput.timeIsNew = gpsData.timeIsNew;
```
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp]: Could use memcpy here:

https://en.cppreference.com/w/cpp/string/byte/memcpy

Same applies to other Get_SensorData functions

File: Autopilot/SensorFusion/Src/SensorFusion.cpp [hunk 6]
```diff
@@ -560,3 +551,44 @@ Airspeed_Data_t SF_GetRawAirspeed()
...
+    altimeterObj->GetResult(altimeterData);
+
+    Altimeter_Data_t altimeterOutput;
+
+    altimeterOutput.pressure = altimeterData.pressure;
+    altimeterOutput.altitude = altimeterData.altitude;
+    altimeterOutput.temp = altimeterData.temp;
+    altimeterOutput.utcTime = altimeterData.utcTime;
+    altimeterOutput.isDataNew = altimeterData.isDataNew;
+    altimeterOutput.status = altimeterData.status;
```
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp]: https://en.cppreference.com/w/cpp/string/byte/memcpy

File: Autopilot/SensorFusion/Src/SensorFusion.cpp [hunk 7]
```diff
@@ -522,17 +522,7 @@ IMU_Data_t SF_GetRawIMU()
...
-    imuOutput.accz = imuData.accz;
-    imuOutput.gyrx = imuData.gyrx;
-    imuOutput.gyry = imuData.gyry;
-    imuOutput.gyrz = imuData.gyrz;
-    imuOutput.magx = imuData.magx;
-    imuOutput.magy = imuData.magy;
-    imuOutput.magz = imuData.magz;
-    imuOutput.isDataNew = imuData.isDataNew;
-    imuOutput.sensorStatus = imuData.sensorStatus;
+    std:memcpy(&imuOutput, &imuData, sizeof(imuOutput));
```
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp]: I'd prefer if you used `sizeof(STRUCT_YOU_ARE_COPYING)` here instead.

ex. `sizeof(IMU_Data_t)`

File: Autopilot/SensorFusion/Src/SensorFusion.cpp [hunk 8]
```diff
@@ -544,10 +534,7 @@ Airspeed_Data_t SF_GetRawAirspeed()
 
     Airspeed_Data_t airspeedOutput;
 
-    airspeedOutput.airspeed = airspeedData.airspeed;
-    airspeedOutput.utcTime = airspeedData.utcTime;
-    airspeedOutput.isDataNew = airspeedData.isDataNew;
-    airspeedOutput.sensorStatus = airspeedData.sensorStatus;
+    std:memcpy(&airspeedOutput, &airspeedData, sizeof(airspeedOutput));
```
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp]: `sizeof(Airspeed_Data_t)`

File: Autopilot/SensorFusion/Src/SensorFusion.cpp [hunk 9]
```diff
@@ -559,19 +546,7 @@ Gps_Data_t SF_GetRawGPS()
...
-    gpsOutput.longitude = gpsData.longitude;
-    gpsOutput.utcTime = gpsData.utcTime;
-    gpsOutput.groundSpeed = gpsData.groundSpeed;
-    gpsOutput.altitude = gpsData.altitude;
-    gpsOutput.heading = gpsData.heading;
-    gpsOutput.numSatellites = gpsData.numSatellites;
-    gpsOutput.fixStatus = gpsData.fixStatus;
-    gpsOutput.sensorStatus = gpsData.sensorStatus;
-    gpsOutput.timeIsNew = gpsData.timeIsNew;
+    std:memcpy(&gpsOutput, &gpsData, sizeof(gpsOutput));
```
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp]: `sizeof(Gps_Data_t)`

File: Autopilot/SensorFusion/Src/SensorFusion.cpp [hunk 10]
```diff
@@ -583,12 +558,7 @@ Altimeter_Data_t SF_GetRawAltimeter()
 
     Altimeter_Data_t altimeterOutput;
 
-    altimeterOutput.pressure = altimeterData.pressure;
-    altimeterOutput.altitude = altimeterData.altitude;
-    altimeterOutput.temp = altimeterData.temp;
-    altimeterOutput.utcTime = altimeterData.utcTime;
-    altimeterOutput.isDataNew = altimeterData.isDataNew;
-    altimeterOutput.status = altimeterData.status;
+    std:memcpy(&altimeterOutput, &altimeterData, sizeof(altimeterOutput));
```
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp]: `sizeof(Altimeter_Data_t)`

File: Autopilot/SensorFusion/Inc/SensorFusion.hpp [hunk 11]
```diff
@@ -65,8 +64,8 @@ typedef struct
     long double longitude; // 8 Bytes
     float utcTime;     // 4 Bytes. Time in seconds since 00:00 (midnight)
     float groundSpeed; // in m/s
-    short heading; // in degrees. Should be between 0-360 at all times, but using integer just in case
     int altitude; // in m
+    short heading; // in degrees. Should be between 0-360 at all times, but using integer just in case
```
Comment (Pop0097) [inline Autopilot/SensorFusion/Inc/SensorFusion.hpp]: comment does not conform to datatype used. 

Also, the GPS struct uses int for heading. For consistency, we should use `int` here too

**Review by Pop0097 — CHANGES_REQUESTED** [review]
Small comments. Also, please address the changes required in the CMake file

**Review by Pop0097 — APPROVED** [review]
Looks good to me

**Pull request merged**